Add pwndbg.storage and four caching API file_cache_{,with_signature_}for_{text,pickle}

MaskRay opened the pull request:
Add pwndbg.storage and four caching API file_cache_{,with_signature_}for_{text,pickle}

Consider an imaginary file pwndbg/foo.py

```python
@pwndbg.storage.file_cache_for_text('{0}.txt')
def ft(x):
    print(x)
    return x
```

```
@pwndbg.storage.file_cache_for_text('c/{0}-{1}.txt')
def fx(x, y): return x
# fx('a', 'b') will save to ~/.cache/pwndbg/foo/ft/c/a-b.txt
```

The first call of `ft('a')` will save a file in `~/.cache/pwndbg/foo/ft/a.txt` .
A later call of `ft('a')` will load the text from the cache.

```python
@pwndbg.storage.file_cache_with_signature_for_text('{0}')
def fts(x):
    print(x)
    return x
```

The caller should provide an additional argument (signature) to fts.
The first call of `fts('a', signature='sig')` will save a file in `~/.cache/pwndbg/foo/fts/a` .
A call of fts('a', signature='sig') will load the cache.
However, a call of fts('a', signature='different_sig') will recalculate as the signature is different.

MaskRay committed:
Add pwndbg.storage and four caching API file_cache_{,with_signature_}for_{text,pickle}

Consider an imaginary file pwndbg/foo.py

@pwndbg.storage.file_cache_for_text('{0}.txt')
def ft(x):
    print(x)
    return x

The first call of ft('a') will save a file ~/.cache/pwndbg/foo/ft/a.txt .
The basename is generated from '{0}.txt'.format(x)
A later call of ft('a') will load the text from the cache.

The template can also be specified as a callable
@pwndbg.storage.file_cache_for_text(lambda x: x.replace('/', '')+'.txt')

@pwndbg.storage.file_cache_with_signature_for_text('{0}')
def fts(x):
    print(x)
    return x

The caller should provide an extra keyword argument (`signature`) to `fts`.
The first call of fts('a', signature='sig') will save a file ~/.cache/pwndbg/foo/fts/a .
A call of fts('a', signature='sig') will load the cache.
However, a call of fts('a', signature='different_sig') will recalculate as the signature is different.

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,41 @@
+            dir = os.path.join(os.getenv('LOCALAPPDATA'), 'pwndbg', 'Cache')
+        else:
+            dir = os.path.join(os.environ.get('XDG_CACHE_HOME') or os.path.expanduser('~/.cache'), 'pwndbg')
+    if sys.version_info[0] >= 3:
```
There is `pwndbg.compat.python3` for that :).

disconnect3d replied:
Also if there isn't a better way for `os.makedirs` with `exist_ok=True` kwarg, I would just leave the py2 implementation (as it will work in py3).

MaskRay replied:
Thanks. Removed py2 stuff.

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,41 @@
+        if sys.platform == 'darwin':
+            dir = os.path.join(os.path.expanduser('~/Library/Caches'), 'pwndbg')
+        elif sys.platform == 'win32':
+            dir = os.path.join(os.getenv('LOCALAPPDATA'), 'pwndbg', 'Cache')
```
Is this even possible to launch pwndbg on Windows?

MaskRay replied:
I also wonder whether it works on Windows...

disconnect3d replied:
We are tied to Elfs and AUXV, so I am almost sure that no.

The AUXV stuff is the reason why pwndbg isn't working on a bare metal currently (see https://github.com/pwndbg/pwndbg/issues/264).

MaskRay replied:
I can remove the two lines for Windows. Do you have concerns regarding to other parts?

disconnect3d replied:
We probably should add a mechanism to memoize or save/load something in cache in this PR.

Just a note: it would be nice to have an assertion to error on saving two things into same file, not to have weird issues in the future (e.g. assert when we register function to be cached pernamently under already registered key/filename).
This might be tricky considering the fact we might have code that reloads modules (see: https://github.com/pwndbg/pwndbg/pull/267).

Hm.. if it is going to be a decorator it could be implemented so that we would take import path + function name as cache key (or rather file name) - this won't have duplicates.

MaskRay replied:
For the https://github.com/pwndbg/pwndbg/pull/261 case, the arguments of the function also decide the name of the cache key. I am not very clear on how to implement this.

disconnect3d commented:
The decorator can get optional argument which would be a function getting
args and kwargs and returning cache key.

09.05.2017 3:32 AM "Fangrui Song" <notifications@github.com> napisał(a):

> *@MaskRay* commented on this pull request.
> ------------------------------
>
> In pwndbg/storage.py
> <https://github.com/pwndbg/pwndbg/pull/268#discussion_r115391164>:
>
> > +# -*- coding: utf-8 -*-
> +from __future__ import absolute_import
> +from __future__ import division
> +from __future__ import print_function
> +from __future__ import unicode_literals
> +
> +import os
> +import sys
> +
> +def get_cache_dir():
> +    dir = os.environ.get('PWNDBG_CACHE_DIR')
> +    if not dir:
> +        if sys.platform == 'darwin':
> +            dir = os.path.join(os.path.expanduser('~/Library/Caches'), 'pwndbg')
> +        elif sys.platform == 'win32':
> +            dir = os.path.join(os.getenv('LOCALAPPDATA'), 'pwndbg', 'Cache')
>
> For the #261 <https://github.com/pwndbg/pwndbg/pull/261> case, the
> arguments of the function also decide the name of the cache key. I am not
> very clear on how to implement this.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/pwndbg/pwndbg/pull/268#discussion_r115391164>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJi7ClhaYIVew1cg1_DIvpdjSULT0WpPks5r38IzgaJpZM4NUqVo>
> .
>

MaskRay commented:
@disconnect3d Refactored.

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,128 @@
+            makedirs(os.path.dirname(path))
+            with open(path, 'wb') as f:
+                pickle.dump((signature, ret), f)
+
```
We are missing `return ret` here

MaskRay replied:
Done

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,128 @@
+            pass
+        if not valid:
+            ret = function(*args)
+            assert isinstance(ret, str)
```
Probably better to have `isinstance(ret, six.string_types)` here - this would let us cache both binary/ascii strings and unicode one (however I am not sure if below file writes pass with both types).

MaskRay replied:
Done. Yes, you can write a `unicode` to `open(..., 'w')` in Python 2.

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,128 @@
+                lines = f.readlines()
+            ret = ''
+            for i in range(len(lines)):
+                match = regex.search(lines[i])
```
Shouldn't we check for signature only in the first line?

MaskRay replied:
Done

disconnect3d commented:
> The caller should provide an additional argument (signature) to fts.
The first call of fts('a', 'sig') will save a file in ~/.cache/pwndbg/foo/fts/a .
A call of fts('a', 'sig') will load the cache.
However, a call of fts('a', 'different_sig') will recalculate as the signature is different.

Do we have any use case for the `signature` decorator now?

I would rather have `signature` passed as kwarg to the decorator function and not to the function I am decorating.

MaskRay commented:
@disconnect3d I'd like to use the `signature` decorator for `find_compilable_headers` in https://github.com/pwndbg/pwndbg/pull/261 . It is a waste of time to find compilable system header files each time pwndbg starts. Thus the code serializes the list of header files somewhere and recalculate whenever the list of system header files changes (which is unlikely to happen if you do not install new packages).

zachriggle commented:
One of the considerations is that the data would need to be cached
per-architecture. The list of headers that works on amd64 may be different
from arm, for example.


On Tue, May 9, 2017 at 7:11 PM Fangrui Song <notifications@github.com>
wrote:

> @disconnect3d <https://github.com/disconnect3d> I'd like to use the
> signature decorator for find_compilable_headers in #261
> <https://github.com/pwndbg/pwndbg/pull/261> . It is a waste of time to
> find compilable system header files each time pwndbg starts. Thus the code
> serializes the list of header files somewhere and recalculate whenever the
> list of system header files changes (which is unlikely to happen if you do
> not install new packages).
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pwndbg/pwndbg/pull/268#issuecomment-300336754>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAG0GEqIKqOQFaEjRLJAgqxX8y2hWlbCks5r4QCcgaJpZM4NUqVo>
> .
>

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,130 @@
+            with open(path, 'rb') as f:
+                (saved_signature, ret) = pickle.load(f)
+                if saved_signature == signature:
+                    valid = True
```
Just `valid = saved_signature == signature`

zachriggle replied:
^ Meh

disconnect3d replied:
Yeah, the condition is more explicit/visible

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,130 @@
+                if match:
+                    saved_signature = match.group(1)
+                    if saved_signature == signature:
+                        ret = ''.join(lines[1:])
```
Those are just lines from the cached return, right? If so, we should probably join it using `'\n'` string, so the newlines are back there.

MaskRay replied:
`readlines` does not strip the newlines

zachriggle commented:
One of the issues with the current implementation is what happens when `f(a, b)` is cached and `a=="../../../../../etc/passwd"` and similar.

MaskRay commented:
Added a simple foolproof check.

Changed to 
```python
        @functools.wraps(function)
        def wrapper(*args, **kwargs):
            if isinstance(template, past.builtins.basestring):
                path = template.format(*args, **kwargs)
            else:
                kwargs1 = kwargs.copy()
                kwargs1.pop('signature', None)
                path = template(*args, **kwargs1)
            path = os.path.join(dirpath, path)
            assert os.path.normpath(path).startswith(dirpath+'/'), 'Please defend path traversal in the template'
            return fn(path, function, args, kwargs)
```

disconnect3d commented:
Hmm... is that ready? Do we still miss the things pointed out by @zachriggle ?
Could rebase it to latest dev @MaskRay?

MaskRay commented:
one year passed and now I'm heading for another Defcon.. should take some
time to re-learn pending..

On Tue, Jul 10, 2018, 13:16 Disconnect3d <notifications@github.com> wrote:

> Hmm... is that ready? Do we still miss the things pointed out by
> @zachriggle <https://github.com/zachriggle> ?
> Could rebase it to latest dev @MaskRay <https://github.com/MaskRay>?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pwndbg/pwndbg/pull/268#issuecomment-403951805>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZaQntrWODION1vgaKbLhVGaRFKJnmqks5uFQuygaJpZM4NUqVo>
> .
>

disconnect3d commented:
Closing this as it is too old :(
